drivers: frequency: cf_axi_dds: Add dacfifo singleshot output support

Yamakaja opened the pull request:
This commit adds support for a new single-shot mode of util_dacfifo.

Note: This feature requires HDL support and is only enabled if the
single-shot-output-gpios attribute is provided in the device tree!

Signed-off-by: David Winter <david.winter@analog.com>

nunojsa commented on drivers/iio/frequency/cf_axi_dds.c:
```diff
@@ -1532,6 +1548,50 @@ static const struct file_operations cf_axi_dds_debugfs_fops = {
+	.open = simple_open,
+	.read = cf_axi_dds_debugfs_read_single_shot,
+	.write = cf_axi_dds_debugfs_write_single_shot,
+};
```
You can use `DEFINE_DEBUGFS_ATTRIBUTE()`...

nunojsa commented on drivers/iio/frequency/cf_axi_dds.c:
```diff
@@ -2251,6 +2311,13 @@ static int cf_axi_dds_probe(struct platform_device *pdev)
+	if (!IS_ERR(st->pl_single_shot_output_gpio) && iio_get_debugfs_dentry(indio_dev)) {
+		debugfs_create_file("pl_single_shot_output", 0644,
+				    iio_get_debugfs_dentry(indio_dev),
+				    indio_dev, &cf_axi_dds_debugfs_fops_single_shot);
```
when using `DEFINE_DEBUGFS_ATTRIBUTE()`, you can create the debugfs file with `debugfs_create_file_unsafe()`

nunojsa commented on drivers/iio/frequency/cf_axi_dds.c:
```diff
@@ -2251,6 +2311,13 @@ static int cf_axi_dds_probe(struct platform_device *pdev)
 				    iio_get_debugfs_dentry(indio_dev),
 				    indio_dev, &cf_axi_dds_debugfs_fops);
 
+	st->pl_single_shot_output_gpio = devm_gpiod_get(&pdev->dev, "single-shot-output", GPIOD_ASIS);
```
This looks like we should use `devm_gpiod_get_optional()`... It will return NULL if the gpio is not given in the devicetree. Then, on `cf_axi_dds_dma_fifo_single_shot()` you do `if (gpio != NULL)`

nunojsa commented on drivers/iio/frequency/cf_axi_dds.c:
```diff
@@ -1532,6 +1548,50 @@ static const struct file_operations cf_axi_dds_debugfs_fops = {
+
+	count = min_t(size_t, count, (sizeof(buf)-1));
+	if (copy_from_user(p, userbuf, count))
+		return -EFAULT;
```
This looks odd. Do we really need to call `copy_from_user()`? If we are passing a boolean `strtobool()` should be enough...

nunojsa commented on drivers/iio/frequency/cf_axi_dds.c:
```diff
@@ -209,6 +217,14 @@ static int cf_axi_dds_signed_mag_fmt_to_iio(unsigned int val, int *r_val,
 	return IIO_VAL_INT_PLUS_MICRO;
 }
 
+int cf_axi_dds_pl_ddr_fifo_ctrl_single_shot(struct cf_axi_dds_state *st, bool enable)
```
`static int`?

nunojsa commented on drivers/iio/frequency/cf_axi_dds.c:
```diff
@@ -113,6 +115,12 @@ bool cf_axi_dds_dma_fifo_en(struct cf_axi_dds_state *st)
+{
+	return st->pl_dma_fifo_single_shot;
+}
+EXPORT_SYMBOL(cf_axi_dds_dma_fifo_single_shot);
```
I would say to remove this function from this PR. I would prefer to add this function when we actually have users for it...

nunojsa reviewed (commented):
Looks good. Some minor comments. Also take a look at the checkpatch error:

https://dev.azure.com/AnalogDevices/OpenSource/_build/results?buildId=2722&view=logs&j=8816ca86-faec-5cfe-601f-19d1b3f9ab6b&t=2446816d-e5a7-5cd4-afcb-81c090218db4&l=146

Yamakaja commented:
Hey @nunojsa, thanks for the feedback!

My work here is pretty much a small modification of what has previously been implemented for plddrbypass, including quirks like the missing `static`.
https://github.com/analogdevicesinc/linux/blob/67ccc88c48d3821550b2bf0feee2e6ae4b0bcf4b/drivers/iio/frequency/cf_axi_dds.c#L228-L246
Should i just rework both attributes accordingly? (Possibly in a separate commit)

nunojsa commented:
> Should i just rework both attributes accordingly? (Possibly in a separate commit)

I would maybe even say on a different PR (for plddrbypass attr)? Anyways, I will leave that up to you. But if you do the refactor in this PR, it definitely needs to go in its own commit...

Yamakaja commented:
V2:

* Switched both attributes to `DEFINE_DEBUGFS_ATTRIBUTE` / `debugfs_create_file_unsafe()`.
* Removed exported getter
* Switched to `devm_gpiod_get_optional()`

nunojsa commented on drivers/iio/frequency/cf_axi_dds.c:
```diff
@@ -209,6 +211,14 @@ static int cf_axi_dds_signed_mag_fmt_to_iio(unsigned int val, int *r_val,
 
+static int cf_axi_dds_pl_ddr_fifo_ctrl_single_shot(struct cf_axi_dds_state *st, bool enable)
+{
+	if (st->pl_single_shot_output_gpio == NULL)
```
nit: same here about the preferred style

nunojsa commented on drivers/iio/frequency/cf_axi_dds.c:
```diff
@@ -1516,6 +1529,31 @@ DEFINE_DEBUGFS_ATTRIBUTE(cf_axi_dds_debugfs_fifo_en_fops,
+	if (!(val == 0 || val == 1))
+		return -EINVAL;
+
+	st->pl_dma_fifo_single_shot = val;
```
same suggestion as before...

nunojsa commented on drivers/iio/frequency/cf_axi_dds.c:
```diff
@@ -1489,6 +1499,9 @@ static const struct iio_info cf_axi_dds_info = {
 
+
+
+
```
just one new line...

nunojsa commented on drivers/iio/frequency/cf_axi_dds.c:
```diff
@@ -1516,6 +1529,31 @@ DEFINE_DEBUGFS_ATTRIBUTE(cf_axi_dds_debugfs_fifo_en_fops,
+static int cf_axi_dds_debugfs_single_shot_get(void *data, u64 *val)
+{
+	struct iio_dev *indio_dev = data;
+	struct cf_axi_dds_state *st = iio_priv(indio_dev);
```
new line here...

nunojsa commented on arch/arm64/boot/dts/xilinx/zynqmp-zcu102-rev10-ad9081-m8-l4.dts:
```diff
@@ -355,3 +355,7 @@
+
+&axi_ad9081_core_tx {
+	single-shot-output-gpios = <&gpio 139 0>;
+};
```
the dts change should be in it's own commit... typically the subject goes like `arch: arm64: dts: ...`

nunojsa reviewed (commented):
Looks good. Just some minor comments and I think we are good to go...

nunojsa reviewed (approved):
There's a checkpatch warning here: https://dev.azure.com/AnalogDevices/OpenSource/_build/results?buildId=2785&view=logs&j=8816ca86-faec-5cfe-601f-19d1b3f9ab6b&t=2446816d-e5a7-5cd4-afcb-81c090218db4&l=160

I will leave it up to you for fixing it. Anyways, for the future, it is a good practice to run checkpatch locally before sending the PR (so, you would have avoided the CI complain :wink: )...

Yamakaja commented:
Alright, next round:

V3:
* Various style fixes
* Broke out dts changes into separate commit
